algebird-core/src/main/scala/com/twitter/algebird/macros/Cuber.scala

```diff
@@ -0,0 +1,81 @@
+package com.twitter.algebird.macros
+
+import scala.language.experimental.{ macros => sMacros }
+import scala.reflect.macros.Context
+import scala.reflect.runtime.universe._
+
+// import com.twitter.algebird._
+
+trait Cuber[I] {
+  type K
+  def apply(in: I): TraversableOnce[K]
+}
+
+trait Roller[I] {
+  type K
+  def apply(in: I): TraversableOnce[K]
+}
+
+object Cuber {
+  def cuber[T]: Cuber[T] = macro cuberImpl[T]
```

Comment:
The cube function in Pig and SQL doesn't seem to have aggregation built in -- it only generates the tuples, and then you manually do a groupby/aggregation/etc afterwards.

Maybe we should reduce the number of things that the `cube` function does to follow that style -- either 
1) only produce the tuples (don't group or aggregate) i.e. return a `TraversableOnce[c.K, TraversableOnce[V]]`
2) produce the tuples and group them (don't aggregate), i.e. return a `Map[c.K, TraversableOnce[V]]`

Then the user can decide what they want to do with the resulting sequence or map.

(This also seems more in line with how we were planning to implement this in Scalding -- TypedPipe#cube would return a TypedPipe[(c.K, V)].)

The method that you mentioned above is definitely useful -- we could call it `cubeAggregate` or something like that to indicate that it's doing more than just cubing the records.